README.md
# trimperson
TRIM API wrapper

## Installation

```
npm install --save mediasuitenz/trimperson#v1.0.0
```

## Usage
### setup module:
```js
var trim = require('trim')({
  url: 'http://path/to/trim/api',
  token: 'trim-api-security-token'
})
```


### TRIM data

Data returned from find looks like:
```json
{
  "containerNo": "BC140111",
  "subContainers": [
    {
      "containerNo": "BC140111#05",
      "subContainers": [

      ],
      "records": [

      ]
    },
    {
      "containerNo": "BC140111#04",
      "subContainers": [

      ],
      "records": [
        {
          "recordNo": "1448586",
          "hyperlink": "http://emap.marlborough.govt.nz/trim/api/trim/get?id=1448586",
          "recordTitle": "2014-2-24 - Inspection - Jetties - gla        - Pass",
          "recordType": "pdf",
          "recordSize": 48519,
          "dateCreated": "2014-02-24T11:56:43",
          "dateRegistered": "2014-02-24T11:56:43",
          "dateFinalized": "2014-02-24T11:56:43",
          "author": ""
        },
        {
          "recordNo": "1448587",
          "hyperlink": "http://emap.marlborough.govt.nz/trim/api/trim/get?id=1448587",
          "recordTitle": "2014-2-24 - Inspection - Veneer Half High - gla        - Pass",
          "recordType": "pdf",
          "recordSize": 48816,
          "dateCreated": "2014-02-24T11:56:47",
          "dateRegistered": "2014-02-24T11:56:47",
          "dateFinalized": "2014-02-24T11:56:47",
          "author": ""
        }
      ]
    },
    {
      "containerNo": "BC140111#03",
      "subContainers": [

      ],
      "records": [
        {
          "recordNo": "13228764",
          "hyperlink": "http://emap.marlborough.govt.nz/trim/api/trim/get?id=13228764",
          "recordTitle": "Issued Building Consent - Building Act 2004 - Field Sheet",
          "recordType": "pdf",
          "recordSize": 77824,
          "dateCreated": "2013-08-12T11:08:36",
          "dateRegistered": "2013-08-12T11:08:38",
          "dateFinalized": "2013-08-15T16:37:31",
          "author": "Cheryl Ewan"
        },
        {
          "recordNo": "13246844",
          "hyperlink": "http://emap.marlborough.govt.nz/trim/api/trim/get?id=13246844",
          "recordTitle": "Engineer involvement in inspections",
          "recordType": "pdf",
          "recordSize": 241152,
          "dateCreated": "2013-08-22T12:29:07",
          "dateRegistered": "2013-08-22T12:29:10",
          "dateFinalized": "2013-08-22T12:40:04",
          "author": "Cheryl Ewan"
        }
      ]
    },
    {
      "containerNo": "BC140111#02",
      "subContainers": [

      ],
      "records": [
        {
          "recordNo": "1448229",
          "hyperlink": "http://emap.marlborough.govt.nz/trim/api/trim/get?id=1448229",
          "recordTitle": "Drainage Plan - TEST",
          "recordType": "pdf",
          "recordSize": 140203,
          "dateCreated": "2014-02-21T14:07:46",
          "dateRegistered": "2014-02-24T07:16:45",
          "dateFinalized": "2014-02-24T07:17:38",
          "author": ""
        }
      ]
    },
    {
      "containerNo": "BC140111#01",
      "subContainers": [

      ],
      "records": [

      ]
    }
  ],
  "records": [

  ]
}
```

#Development

1. Get the code `git checkout git@github.com:mediasuitenz/trimperson.git`
2. Install dependencies `cd trimperson && npm install`
3. Run tests `npm run test`
4. alternative to (3) run `npm run test:watch` to have tests run whenever a file changes



Changelog.md
# v3.0.0
Introduce better access control.

## Creating Records
- Change createRecord parameters to accept a data object in instead of a bunch of sequential parameters
- createRecord will still accept sequential parameters for backwards compatibility
- createRecord accepts new privacyLevel parameter
- add createPrivateRecord
- add createPublicRecord

## Fetching Containers

- add getPublicContainer

## Creating containers

- Change createContainer parameters to accept a data object instead of a bunch of sequential parameters
- createContainer will still accept sequential parameters for backwards compatibility
- createContainer accepts new privacyLevel parameter
- Add createPublicContainer
- Add createPrivateContainer

## Error handling
The requests now actually handle error responses instead of relying on consumers to
parse and intuit the meaning of responses. A bunch of error conditions have been introduced
that will return error values in the response callbacks.

## Configuring
Remove the old trim.development and therefore mocking.

# v2.2.1
Unrecorded.


specs.js
/* global Given, Then, describe */
var chai = require('chai')
var expect = chai.expect

var nock = require('nock')
var R = require('ramda-extended')

var VALID_TITLE = 'Some Title'
var INVALID_TITLE = null

var VALID_RECORD_NO = 'SomeRecordNo'
var INVALID_RECORD_NO = ''

var VALID_EXTENSION = 'pdf'
var INVALID_EXTENSION = '.pdf'

var VALID_ALTERNATIVE_CONTAINERS = ['SomeContainer']
var INVALID_ALTERNATIVE_CONTAINERS = null

var VALID_CONTAINER = 'oijsdoi'
var INVALID_CONTAINER = null

var VALID_FILEDATA = 'R0lGOD lhCwAOAMQfAP////7+/vj4+Hh4eHd3d/v7+/Dw8HV1dfLy8ubm5vX19e3t7fr'
var INVALID_FILEDATA = ''

var VALID_FOLDER_NAME = 'Some folder'
var INVALID_FOLDER_NAME = ''

describe('When using v2.2.1 or less', () => {
  var trim, mockURL, mockToken

  Given(() => mockURL = 'http://localhost')
  Given(() => mockToken = 'DEADBEEF')

  Given(() => {
    trim = require('./index')({
      url: mockURL,
      token: mockToken,
      testing: true,
      mock: false
    })
  })

  describe('You should be able to call `createRecord` with 5 arguments (include alternativeContainers)', () => {
    var title, containerId, extensionType, fileData, alternativeContainers, createRecordMock

    Given(() => title = 'inside-out')
    Given(() => containerId = 'asdfjk')
    Given(() => extensionType = '.mp4')
    Given(() => alternativeContainers = ['himark'])
    Given(() => fileData = VALID_FILEDATA)

    var errReturn
    var dataReturn
    Given(() => {
      createRecordMock = nock(mockURL)
          .filteringRequestBody(R.always('*'))
          .post('/AddRecordToTrim', '*')
          .query({securityToken: mockToken})
          .reply(201, {RecordNo: VALID_RECORD_NO})
    })
    When((done) => {
      trim.createRecord(title, containerId, extensionType, fileData, alternativeContainers, function (err, data) {
        errReturn = err
        dataReturn = data
        done()
      })
    })
    Then('it should be able to successfully use the function', (done) => {
      expect(errReturn).not.to.exist
      expect(dataReturn).to.exist
      expect(dataReturn).to.be.a('string')
      expect(dataReturn).to.equal(VALID_RECORD_NO)
      createRecordMock.done()
      done()
    })

  })

  describe('You should be able to call `createRecord` with 4 arguments (no alternativeContainers)', () => {
    var title, containerId, extensionType, fileData, createRecordMock

    Given(() => title = VALID_FOLDER_NAME)
    Given(() => containerId = 'asdfjk')
    Given(() => extensionType = '.mp4')
    Given(() => fileData = 'R0lGOD lhCwAOAMQfAP////7+/vj4+Hh4eHd3d/v7+/Dw8HV1dfLy8ubm5vX19e3t7fr')

    var errReturn
    var dataReturn
    Given(() => {
      createRecordMock = nock(mockURL)
          .filteringRequestBody(R.always('*'))
          .post('/AddRecordToTrim', '*')
          .query({securityToken: mockToken})
          .reply(201, {RecordNo: VALID_RECORD_NO})
    })
    When((done) => {
      trim.createRecord(title, containerId, extensionType, fileData, function (err, data) {
        errReturn = err
        dataReturn = data
        done()
      })
    })
    Then('it should be able to successfully use the function', (done) => {
      expect(errReturn).not.to.exist
      expect(dataReturn).to.exist
      expect(dataReturn).to.be.a('string')
      expect(dataReturn).to.equal(VALID_RECORD_NO)
      createRecordMock.done()
      done()
    })

  })

  describe('You should be able to call `createContainer` with 4 arguments (no alternativeContainers)', () => {
    var createContainerMock, folder, parentFolder, privacySetting
    var dataReturn
    var errReturn

    Given(() => folder = VALID_FOLDER_NAME)
    Given(() => parentFolder = 'someParentFolder')
    Given(() => privacySetting = trim.PRIVACY.PUBLIC)
    Given(() => {
      createContainerMock = nock(mockURL)
          .filteringRequestBody(R.always('*'))
          .post('/CreateContainer', '*')
          .query({securityToken: mockToken})
          .reply(201, {
            RecordNo: VALID_FOLDER_NAME
          })
    })
    When((done) => {
      trim.createContainer(folder, privacySetting, parentFolder, function (err, data) {
        errReturn = err
        dataReturn = data
        done()
      })
    })
    Then('it should be able to successfully use the function', (done) => {
      expect(errReturn).not.to.exist
      expect(dataReturn).to.exist
      createContainerMock.done()
      done()
    })
  })

})

describe('When using trimperson api', () => {
  var trim
  var mockURL
  var mockToken

  Given(() => mockURL = 'http://localhost')
  Given(() => mockToken = 'DEADBEEF')

  Given(() => {
    trim = require('./index')({
      url: mockURL,
      token: mockToken,
      testing: true,
      mock: false
    })
  })

  Then('it should have created a trimperson object', (done) => {
    expect(trim).to.be.an('object')
    done()
  })

  Then('Unit test: isValidPrivacyLevel', (done) => {
    expect(trim.isValidPrivacyLevel(trim.PRIVACY.PUBLIC)).to.equal(true)
    expect(trim.isValidPrivacyLevel(trim.PRIVACY.PRIVATE)).to.equal(true)
    expect(trim.isValidPrivacyLevel(3)).to.equal(false)
    done()
  })


  /*
   CONSTANTS
   */
  describe('Privacy constants should be available', () => {
    Then('`PRIVACY_LEVELS` should exist', (done) => {
      expect(trim.PRIVACY).to.be.an('object')
      done()
    })
    Then('`PRIVACY_LEVELS` should have the correct PUBLIC value', (done) => {
      expect(trim.PRIVACY.PUBLIC).to.equal(1)
      done()
    })
    Then('`PRIVACY_LEVELS` should have the correct PRIVATE value', (done) => {
      expect(trim.PRIVACY.PRIVATE).to.equal(2)
      done()
    })
  })

  /*
   CREATE RECORD
   */
  describe('To create a record', () => {

    Then('`createRecord` should be a function', (done) => {
      expect(trim.createRecord).to.be.a('function')
      done()
    })
    describe('Using the `createRecord` function', () => {

      describe('When the request succeeds', () => {
        var data, errReturn, dataReturn, createRecordMock
        Given(() => data = {
          title: VALID_TITLE,
          container: VALID_CONTAINER,
          extension: VALID_EXTENSION,
          fileData: VALID_FILEDATA,
          alternativeContainers: VALID_ALTERNATIVE_CONTAINERS,
          privacyLevel: trim.PRIVACY.PUBLIC
        })
        Given(() => {
          createRecordMock = nock(mockURL)
              .filteringRequestBody(R.always('*'))
              .post('/AddRecordToTrim', '*')
              .query({securityToken: mockToken})
              .reply(201, {
                RecordNo: '097097234'
              })
        })

        When(done => {
          trim.createRecord(data, (err, data) => {
            errReturn = err
            dataReturn = data
            done()
          })
        })

        Then('It should return the container', (done) => {
          expect(errReturn).not.to.exist
          expect(dataReturn).to.be.a('string')
          expect(dataReturn).to.equal('097097234')
          createRecordMock.done()
          done()
        })
      })

      describe('When the request fails', () => {
        var errReturn, dataReturn, data

        Given(() => data = {
          title: VALID_TITLE,
          container: VALID_CONTAINER,
          extension: VALID_EXTENSION,
          fileData: VALID_FILEDATA,
          alternativeContainers: VALID_ALTERNATIVE_CONTAINERS,
          privacyLevel: trim.PRIVACY.PUBLIC
        })
        Given(() => {
          createRecordMock = nock(mockURL)
              .filteringRequestBody(R.always('*'))
              .post('/AddRecordToTrim', '*')
              .query({securityToken: mockToken})
              .reply(500, {message: 'An error has occurred.'})
        })
        When((done) => {
          trim.createRecord(data, function (err, data) {
            errReturn = err
            dataReturn = data
            done()
          })
        })
        Then('It should return an error', (done) => {
          expect(errReturn).to.exist
          expect(dataReturn).not.to.exist
          createRecordMock.done()
          done()
        })
      })
    })


  })

  /*
   GET RECORD
   */
  describe('To get a single record', () => {
    Then('`getDocument` should be a function', (done) => {
      expect(trim.getDocument).to.be.a('function')
      done()
    })

    describe('When the record exists', () => {
      var documentId, getDocumentMock, dataReturn, errReturn

      Given(() => documentId = 'someIDthatPointsToADocument')
      Given(() => {
        getDocumentMock = nock(mockURL)
            .get('/get')
            .query({id: documentId, securityToken: mockToken})
            .reply(200, VALID_FILEDATA)
      })

      When((done) => {
        trim.getDocument(documentId, function (err, data) {
          dataReturn = data
          errReturn = err
          done()
        })
      })

      Then('It should return a document', (done) => {
        expect(errReturn).not.to.exist
        expect(dataReturn).to.equal(VALID_FILEDATA)
        getDocumentMock.done()
        done()
      })
    })
    describe('When the record doesnt exist', () => {
      var documentId, getDocumentMock, dataReturn, errReturn

      Given(() => documentId = 'someIDthatPointsToADocument')
      Given(() => {
        getDocumentMock = nock(mockURL)
            .get('/get')
            .query({id: documentId, securityToken: mockToken})
            .reply(404, '')
      })

      When((done) => {
        trim.getDocument(documentId, function (err, data) {
          dataReturn = data
          errReturn = err
          done()
        })
      })

      Then('It should return an error', (done) => {
        expect(errReturn).to.exist
        expect(dataReturn).not.to.exist
        getDocumentMock.done()
        done()
      })
    })
  })

  /*
   CREATE CONTAINER
   */
  describe('To create a container', () => {
    Then('`createContainer` should be a function', (done) => {
      expect(trim.createContainer).to.be.a('function')
      done()
    })

    describe('If all inputs are valid', () => {
      var createContainerMock, dataReturn, errReturn, data

      Given(() => data = {
        folderName: VALID_FOLDER_NAME,
        privacyLevel: trim.PRIVACY.PUBLIC
      })
      Given(() => {
        createContainerMock = nock(mockURL)
            .filteringRequestBody(R.always('*'))
            .post('/CreateContainer', '*')
            .query({securityToken: mockToken})
            .reply(201, {
              RecordNo: VALID_FOLDER_NAME
            })
      })
      When(done => {
        trim.createContainer(data, (err, body) => {
          errReturn = err
          dataReturn = body
          done()
        })
      })
      Then('The container is created', done => {
        expect(errReturn).not.to.exist
        expect(dataReturn).to.be.an('object')
        expect(dataReturn.RecordNo).to.equal(VALID_FOLDER_NAME)
        createContainerMock.done()
        done()
      })
    })

    describe('If no privacyLevel is specified', () => {
      var createContainerMock, dataReturn, errReturn, data

      Given(() => data = {
        folderName: VALID_FOLDER_NAME
      })
      Given(() => {
        createContainerMock = nock(mockURL)
            .filteringRequestBody(R.always('*'))
            .post('/CreateContainer', '*')
            .query({securityToken: mockToken})
            .reply(201, {
              RecordNo: VALID_FOLDER_NAME
            })
      })
      When(done => {
        trim.createContainer(data, (err, body) => {
          errReturn = err
          dataReturn = body
          done()
        })
      })
      Then('The container is created', done => {
        expect(errReturn).not.to.exist
        expect(dataReturn).to.be.an('object')
        expect(dataReturn.RecordNo).to.equal(VALID_FOLDER_NAME)
        createContainerMock.done()
        done()
      })
    })

    describe('No folder name is provided', () => {
      var createContainerMock, dataReturn, errReturn, data

      Given(() => data = {
        folderName: ''
      })
      Given(() => {
        createContainerMock = nock(mockURL)
            .filteringRequestBody(R.always('*'))
            .post('/CreateContainer', '*')
            .query({securityToken: mockToken})
            .reply(500, {
              RecordNo: null
            })
      })
      When(done => {
        trim.createContainer(data, (err, body) => {
          errReturn = err
          dataReturn = body
          done()
        })
      })
      Then('The container is not created', done => {
        expect(errReturn).to.exist
        expect(dataReturn).not.to.exist
        createContainerMock.done()
        done()
      })
    })

    describe('With escalated permissions and an invalid token', () => {
      var createContainerMock, dataReturn, errReturn, data

      Given(() => data = {
        folderName: VALID_FOLDER_NAME,
        privacyLevel: trim.PRIVACY.PUBLIC
      })
      Given(() => {
        createContainerMock = nock(mockURL)
            .filteringRequestBody(R.always('*'))
            .post('/CreateContainer', '*')
            .query({securityToken: mockToken})
            .reply(401, {
              message: 'Authorization has been denied for this request.'
            })
      })
      When(done => {
        trim.createContainer(data, (err, body) => {
          errReturn = err
          dataReturn = body
          done()
        })
      })
      Then('An error is returned', done => {
        expect(errReturn).to.exist
        expect(dataReturn).not.to.exist
        createContainerMock.done()
        done()
      })
    })

    describe('A null `RecordNo` is returned', () => {
      var createContainerMock, dataReturn, errReturn, data

      Given(() => data = {
        folderName: VALID_FOLDER_NAME,
        privacyLevel: trim.PRIVACY.PUBLIC
      })
      Given(() => {
        createContainerMock = nock(mockURL)
            .filteringRequestBody(R.always('*'))
            .post('/CreateContainer', '*')
            .query({securityToken: mockToken})
            .reply(201, {
              RecordNo: null
            })
      })
      When(done => {
        trim.createContainer(data, (err, body) => {
          errReturn = err
          dataReturn = body
          done()
        })
      })
      Then('An error is returned', done => {
        expect(errReturn).to.exist
        expect(dataReturn).not.to.exist
        createContainerMock.done()
        done()
      })
    })

    describe('The container already exists', () => {
      var createContainerMock, dataReturn, errReturn, data

      Given(() => data = {
        folderName: VALID_FOLDER_NAME,
        privacyLevel: trim.PRIVACY.PUBLIC
      })
      Given(() => {
        createContainerMock = nock(mockURL)
            .filteringRequestBody(R.always('*'))
            .post('/CreateContainer', '*')
            .query({securityToken: mockToken})
            .reply(204)
      })
      When(done => {
        trim.createContainer(data, (err, body) => {
          errReturn = err
          dataReturn = body
          done()
        })
      })
      Then('An error is returned', done => {
        expect(errReturn).not.to.exist
        expect(dataReturn).to.exist
        expect(dataReturn.RecordNo).to.equal(VALID_FOLDER_NAME)
        createContainerMock.done()
        done()
      })
    })
  })

  /*
   GET CONTAINER
   */
  describe('To get a container', () => {
    Then('`getContainer` should be a function', (done) => {
      expect(trim.getContainer).to.be.a('function')
      done()
    })

    describe('If the public container is requested', () => {
      var getContainerMock
      Given(() => {
        getContainerMock = nock(mockURL)
            .get('/GetContainer')
            .query({trimid: VALID_CONTAINER, securityToken: mockToken})
            .reply(200)
      })
      When(done => {
        trim.getContainer(VALID_CONTAINER, trim.PRIVACY.PUBLIC, () => done())
      })
      Then('`/getContainer` is called', done => {
        getContainerMock.done()
        done()
      })
    })

    describe('If the private container is requested', () => {
      var getContainerMock
      Given(() => {
        getContainerMock = nock(mockURL)
            .get('/getPrivateContainer')
            .query({trimid: VALID_CONTAINER, securityToken: mockToken})
            .reply(200)
      })
      When(done => {
        trim.getContainer(VALID_CONTAINER, trim.PRIVACY.PRIVATE, () => done())
      })
      Then('`/getPrivateContainer` is called', done => {
        getContainerMock.done()
        done()
      })
    })

    describe('If no privacy level is supplied', () => {
      var getContainerMock
      Given(() => {
        getContainerMock = nock(mockURL)
            .get('/GetContainer')
            .query({trimid: VALID_CONTAINER, securityToken: mockToken})
            .reply(200)
      })
      When(done => {
        trim.getContainer(VALID_CONTAINER, () => done())
      })
      Then('The public container is requested', done => {
        getContainerMock.done()
        done()
      })
    })

    describe('With escalated permissions and an invalid token', () => {
      var getContainerMock, dataReturn, errReturn
      Given(() => {
        getContainerMock = nock(mockURL)
            .get('/getPrivateContainer')
            .query({trimid: VALID_CONTAINER, securityToken: mockToken})
            .reply(401, {
              message: 'Authorization has been denied for this request.'
            })
      })
      When(done => {
        trim.getContainer(VALID_CONTAINER, trim.PRIVACY.PRIVATE, (err, data) => {
          errReturn = err
          dataReturn = data
          done()
        })
      })
      Then('It should return an error', done => {
        expect(errReturn).to.exist
        expect(dataReturn).not.to.exist
        getContainerMock.done()
        done()
      })
    })

    describe('When the container exists', () => {
      var getContainerMock, dataReturn, errReturn

      Given(() => {
        getContainerMock = nock(mockURL)
            .get('/GetContainer')
            .query({trimid: VALID_CONTAINER, securityToken: mockToken})
            .reply(200, {
              containerNo: 'someid',
              subContainers: []
            })
      })

      When((done) => {
        trim.getContainer(VALID_CONTAINER, trim.PRIVACY.PUBLIC, (err, data) => {
          dataReturn = data
          errReturn = err
          done()
        })
      })

      Then('It should return the container', (done) => {
        expect(errReturn).not.to.exist
        expect(dataReturn).to.be.an('object')
        expect(dataReturn.containerNo).to.equal('someid')
        expect(dataReturn.subContainers).to.be.a('array')
        getContainerMock.done()
        done()
      })

    })
    describe('When the container doesnt exist', () => {
      var getContainerMock, dataReturn, errReturn

      Given(() => {
        getContainerMock = nock(mockURL)
            .get('/GetContainer')
            .query({trimid: VALID_CONTAINER, securityToken: mockToken})
            .reply(200, null)
      })

      When((done) => {
        trim.getContainer(VALID_CONTAINER, trim.PRIVACY.PUBLIC, (err, data) => {
          dataReturn = data
          errReturn = err
          done()
        })
      })

      Then('It should return an error', (done) => {
        expect(errReturn).to.exist
        expect(dataReturn).not.to.exist
        getContainerMock.done()
        done()
      })
    })

  })
})


trim.development.js
var assert = require('assert');
var url, token;

function createRecord (title, container, extension, fileData, alternativeContainers, callback) {
  callback(null, "13228764")
}

function getDocument (trimId, callback) {
  callback(null, null);
}

function getContainer (trimId, callback) {
  return callback(null, require('./mock/getContainer.json'));
}

function createContainer (folderName, privacy, parentFolder, callback) {
  callback(null, {}); // dunno what this is supposed to be.
}

function getPrivateContainer(trimId, callback) {
  return callback(null, require('./mock/getContainer.json'));
}

module.exports = function (apiUrl, apiToken) {
  assert(typeof apiUrl === 'string', 'Argument 1 to instantiate the TRIM wrapper must be a valid url.')
  assert(typeof apiToken === 'string', 'Argument 2 to instantiate the TRIM wrapper must be a valid apiToken.')

  url = apiUrl;
  token = apiToken;

  return {
    getContainer: getContainer,
    getDocument: getDocument,
    createContainer: createContainer,
    createRecord: createRecord,
    getPrivateContainer: getPrivateContainer
  }
}


trim.production.js
var request = require('request')
var assert = require('assert')
var debug = require('debug')('trim')
var R = require('ramda-extended')
var url
var token
var testing

var PRIVACY = {
  PUBLIC: 1,
  PRIVATE: 2
}

var isValidPrivacyLevel = R.contains(R.__, R.values(PRIVACY))

function deprecationWarning(warning) {
  if (!testing) {
    console.log('DEPRECATION WARNING: ', warning)
  }
}

/**
 * Define the callback from the `createRecord` method
 * @callback createRecordCallback
 * @param {Error}  error
 * @param {String} recordNo The TRIM RecordNo of the new document, if created.
 */

/**
 * Uploads a Document to TRIM
 * @param {Object}    data
 * @param {String}    data.title
 * @param {String}    data.container Primary container to upload to
 * @param {String}    data.extension Omit the period, e.g. "pdf" or "png" but not ".png"
 * @param {String}    data.fileData Base64 String without the prefix
 * @param {String[]}  data.alternativeContainers A list of additional TRIM containers to upload to
 * @param {Boolean}   data.privacyLevel public=1, private=2 -- higher numbers escalate privacy level
 * @param {createRecordCallback}  callback
 */
function createRecord(data, callback) {
  if (arguments.length === 5) {
    // Compatibility with usage before the alternativeContainers parameter was added
    deprecationWarning('The createRecord arguments have changed to accept a single data parameter instead of multiple')
    data = {
      title: arguments[0],
      container: arguments[1],
      extension: arguments[2],
      fileData: arguments[3],
      alternativeContainers: [],
      privacyLevel: 1
    }
    callback = arguments[4]
  } else if (arguments.length === 6) {
    // Compatibility with usage after the alternativeContainers parameter was added
    deprecationWarning('The createRecord arguments have changed to accept a single data parameter instead of multiple')
    data = {
      title: arguments[0],
      container: arguments[1],
      extension: arguments[2],
      fileData: arguments[3],
      alternativeContainers: arguments[4]
    }
    callback = arguments[5]
  }
  if (data.privacyLevel != null && !isValidPrivacyLevel(data.privacyLevel)) {
    return callback(new Error('Invalid privacy level: ' + data.privacyLevel))
  }

  var options = {
    url: url + '/AddRecordToTrim?securityToken=' + token,
    method: 'post',
    body: {
      Title: data.title,
      Container: data.container,
      RecordExtension: data.extension,
      AlternativeContainers: data.alternativeContainers || [],
      Record: data.fileData,
      Privacy: data.privacyLevel || PRIVACY.PUBLIC
    },
    json: true
  }
  request(options, function (err, res, responseBody) {
    if (err) return callback(err)

    var trimRecordNo = responseBody.RecordNo

    if (!trimRecordNo) {
      debug('Invalid response %j', responseBody)
      return callback(new Error('Error uploading document to TRIM: Missing RecordNo response. ' + responseBody.message))
    }
    debug('Created record with recordNo %s', trimRecordNo)
    return callback(null, trimRecordNo)
  })
}

/**
 * Get the actual document, not the TRIM record
 * @param trimId
 * @param callback
 */
function getDocument (trimId, callback) {
  var options = {
    url: url + '/get?id=' + trimId + '&securityToken=' + token
  }
  request.get(options, function (err, res, responseBody) {
    if (res.statusCode === 404) {
      return callback(new Error('Not found'))
    }
    callback(err, responseBody)
  })
}

/**
 * This callback returns a container
 * @callback containerCallback
 * @param {Error}     error
 * @param {Object}    container
 * @param {String}    container.containerNo
 * @param {Object[]}  container.subcontainers
 * @param {Object[]}  container.records
 */

/**
 *
 * @param {String} method
 * @param {String} trimId
 * @param {containerCallback} callback
 */
function getContainerUsingMethod(method, trimId, callback) {
  var options = {
    url: url + '/' + method + '?trimid=' + trimId + '&securityToken=' + token,
    json: true
  }
  debug('GET %s', options.url)
  request.get(options, function (err, res, responseBody) {
    if (err) {
      callback(err)
    } else if (res.statusCode === 401) {
      callback(new Error('Unauthorized'))
    } else if (responseBody == null) {
      callback(new Error('Not Found'))
    } else {
      callback(err, responseBody)
    }
  })
}

/**
 *
 * @param {String} trimId
 * @param {Number} [privacyLevel]
 * @param {containerCallback} callback
 * @return {*}
 */
function getContainer (trimId, privacyLevel, callback) {
  if (arguments.length === 2) {
    callback = privacyLevel
    privacyLevel = PRIVACY.PUBLIC
  }
  var method
  switch (privacyLevel) {
    case PRIVACY.PUBLIC:
      method = 'GetContainer'
      break
    case  PRIVACY.PRIVATE:
      method = 'getPrivateContainer'
      break
    default:
      return callback(new Error('Invalid privacyLevel: ' + privacyLevel))
  }
  return getContainerUsingMethod(method, trimId, callback)
}

function getPrivateContainer (trimId, callback) {
  deprecationWarning('getPrivateContainer is deprecated. Use getContainer and pass a privacy level instead.')
  return getContainer(trimId, PRIVACY.PRIVATE, callback)
}


/**
 * @callback createContainerCallback
 * @param {Error} error
 * @param {Object} response
 * @param {String} response.RecordNo
 */

/**
 * @param {Object} data
 * @param {String} data.folderName the RecordNo
 * @param {Number} data.privacyLevel
 * @param {String} [data.parentFolder]
 * @param {String} [data.title] Display title? Not actually sure what this does
 * @param {createContainerCallback} callback
 */
function createContainer (data, callback) {
  if (arguments.length === 4) {
    // Compatibility with old api
    data = {
      folderName: arguments[0],
      privacyLevel: arguments[1],
      parentFolder: arguments[2]
    }
    callback = arguments[3]
  }
  var body = {
    RecordNo: data.folderName,
    Title: data.title || data.folderName,
    Privacy: data.privacyLevel || PRIVACY.PUBLIC,
    ParentFolder: data.parentFolder
  }
  // parentFolder currently returns a 500 error
  var options = {
    url: url + '/CreateContainer?securityToken=' + token,
    json: body
  }

  request.post(options, function (err, res, responseBody) {
    if (err) {
      callback(err)
    } else if (res.statusCode === 401) {
      callback(new Error('Unauthorized'))
    } else if (parseInt(res.statusCode / 100, 10) === 5) {
      callback(new Error('Server Error'))
    } else if (responseBody && !responseBody.RecordNo) {
      callback(new Error('Error creating container'))
    } else if (res.statusCode === 204) {
      // if no response body, then the container already existed... what do we do?
      // TODO: Don't know if this is the best course of action...
      callback(null, {RecordNo: data.folderName})
    } else {
      callback(null, responseBody)
    }
  })
}

/**
 *
 * @param {Object} options configuration options
 * @param {String} options.url Base TRIM url
 * @param {String} options.token TRIM securityToken
 * @param {Boolean} options.debug If true, then require `request-debug`
 * @param {Boolean} options.testing Hide deprecation warnings
 * @return {{getContainer: getContainer, getDocument: getDocument, createContainer: createContainer, createRecord:
 *     createRecord}}
 */
module.exports = function (options) {

  assert(typeof options.url === 'string', 'TRIM wrapper must have a valid `url` argument.')
  assert(typeof options.token === 'string', 'TRIM wrapper must have a valid `token` argument.')

  url = options.url
  token = options.token
  testing = options.testing
  if (!!options.debug)
    require('request-debug')(request)

  return {
    PRIVACY: PRIVACY,
    isValidPrivacyLevel: isValidPrivacyLevel,

    getDocument: getDocument,
    createRecord: createRecord,

    getContainer: getContainer,
    createContainer: createContainer,

    // deprecations
    getPrivateContainer: getPrivateContainer

  }
}
